pinot-tools/src/main/java/com/linkedin/pinot/tools/ClusterStateVerifier.java

```diff
@@ -0,0 +1,146 @@
+
+import com.linkedin.pinot.controller.ControllerConf;
+import com.linkedin.pinot.controller.helix.core.PinotHelixResourceManager;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.IdealState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class ClusterStateVerifier {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ClusterStateVerifier.class);
+  private String _clusterName;
+  private PinotHelixResourceManager _helixResourceManager;
+  private String _zkAddress;
+
+  public ClusterStateVerifier(String zkAddress, String clusterName) {
+    _zkAddress = zkAddress;
+    _clusterName = clusterName;
+    _helixResourceManager =  new PinotHelixResourceManager(getControllerConf());
+    _helixResourceManager.start();
+  }
+
+  private ControllerConf getControllerConf() {
+    ControllerConf conf = new ControllerConf();
+    conf.setHelixClusterName(_clusterName);
+    conf.setZkStr(_zkAddress);
+    conf.setTenantIsolationEnabled(false);
+    conf.setControllerVipHost("localhost");
+    conf.setControllerVipProtocol("http");
+    return conf;
+  }
+
+  /**
+   * return 0 if all the tables are stable
+   * @return
+   */
+  private int isClusterStable(final List<String> tableNames, final long timeoutSec) {
+
+    final ExecutorService executor = Executors.newSingleThreadExecutor();
+    final Future<Integer> future = executor.submit(new Callable<Integer>() {
+      @Override
+      public Integer call() {
+        return isClusterStable(tableNames);
+      }
+    });
+
+    Integer isStable = -1;
+    try {
+      isStable = future.get(timeoutSec, TimeUnit.SECONDS);
+    } catch (InterruptedException | ExecutionException ie) {
+      /* Handle the interruption. Or ignore it. */
+      LOGGER.error("Exception occurred: {}", ie);
+    } catch (TimeoutException te) {
+      /* Handle the timeout. */
+      LOGGER.error("Reach timeout! timeoutSec: {}", timeoutSec);
+    } finally {
+      if (!executor.isTerminated()) {
+        executor.shutdownNow();
+      }
+    }
+    return isStable;
+  }
+
+  /**
+   * return 0 if all the tables are stable
+   * @return
+   */
+  private Integer isClusterStable(final List<String> tableNames) {
+    LOGGER.info("Start checking the stability of all the tables...");
+    int isStable = 0;
+    for (String tableName : tableNames) {
+      isStable = isStable(tableName);
+      if (isStable != 0) {
+        LOGGER.error("Table {} is not stable. The cluster is not stable.", tableName);
+        break;
+      }
+    }
+    return isStable;
+  }
+
+  private int isStable(String tableName) {
```

Comment:
You can return a boolean from here. The idea was to exit the command with an exit status of 1 (non-zero) if the cluster is not stable, and 0 if the cluster is. Not sure if our Command hiearchy allows return codes, or do we have to thrown an exception in order to exit with a non-zero code (in Linux)